Allow multiple encryption and signing keys

jonnyshaw89 opened the pull request:
### What is the context of this PR?
Survey Runner is able to accept multiple keys from upstream systems. This allows for multiple systems to integrate and for multiple keys to be supported for each system. Thus making the rotation of keys a zero downtime process.

### How to review 
Add Multiple keys to the secrets and test authenticating with different kid values matching the keys you have added.

collisdigital commented on app/authentication/jwt_decoder.py:
```diff
@@ -22,27 +22,24 @@ class JWTDecryptor(JWERSAOAEPDecryptor):
+                public_key = self._secret_store.get_public_key_by_kid(kid)
+
+                # oddly the python cryptography library needs these as bytes string
+                rrm_public_key_as_bytes = strings.to_bytes(public_key)
```
avoid calling this rrm anymore?

collisdigital commented on app/authentication/jwt_decoder.py:
```diff
@@ -169,6 +172,8 @@ def __check_jwe_protected_header(self, header):
         if header_data.get("enc") != "A256GCM":
             raise InvalidTokenException("Invalid Encoding")
+        if not header_data.get("kid"):
+            raise InvalidTokenException("Missing kid")
```
Need to check with RRM, we may have to support kid as optional initially

collisdigital commented on app/cryptography/jwe_decryption.py:
```diff
@@ -52,19 +48,31 @@ def decrypt(self, token):
+
+        private_key = serialization.load_pem_private_key(
+            self._secret_store.get_private_key_by_kid(kid).encode(),
+            password="digitaleq".encode(),
```
hmmm, hard coded...

jonnyshaw89 replied:
Open question as to wether or not we still want passwords on the private keys?

collisdigital replied:
Seeing as the password is public it looks to offer little value, go-launch-a-survey also doesn't support password protected private keys.

collisdigital commented on app/secrets.py:
```diff
@@ -15,8 +13,52 @@
+        if required_secret not in secrets['secrets']:
             raise Exception("Missing Secret [{}]".format(required_secret))
+
+    # for required_private_key in EXPECTED_PRIVATE_KEYS:
```
Remove

collisdigital commented on app/secrets.py:
```diff
@@ -15,8 +13,52 @@
+            self.public_keys = secrets['public_keys']
+
+    def get_secret_by_name(self, secret_name):
+        if secret_name in self.secrets:
```
should these raise exceptions if not found?

jonnyshaw89 replied:
The validate_required_secrets validates that on startup so the application won't deploy if its missing a secret. If its not required then None seems like a good value to return?

collisdigital replied:
in that case you can just use `self.secrets.get(secret_name)`

collisdigital commented on app/secrets.py:
```diff
@@ -15,8 +13,52 @@
+        if secret_name in self.secrets:
+            return self.secrets[secret_name]
+
+    def get_private_key_by_kid(self, kid):
```
we discussed the concept of key purposes, so you restrict even further, e.g. `get_private_key_by_kid(self, kid, purpose)`

jonnyshaw89 replied:
Yeh, that was next on the list to think about.

collisdigital replied:
👍

collisdigital commented on app/secrets.py:
```diff
@@ -18,5 +16,31 @@
+        raise InvalidTokenException("Invalid Private Key Identifier [{}]".format(kid))
+
+    def get_public_key_by_kid(self, kid):
+        if kid in self.public_keys:
```
same about using `.get()`

collisdigital commented on app/secrets.py:
```diff
@@ -18,5 +16,31 @@
+        raise InvalidTokenException("Invalid Private Key Identifier [{}]".format(kid))
+
+    def get_public_key_by_kid(self, kid):
+        if kid in self.public_keys:
```
do we need to do any validation that `kid` is valid (e.g. is 160 bits etc.)?

jonnyshaw89 replied:
I agree. But, Would need to work out how we identify the keys where we only have one version. e.g. how would we know what public key we would use for submission?

collisdigital replied:
if we had the concept of key types/purposes you could request the key by purpose (and use the first?)

jonnyshaw89 replied:
Or require just 1 to be present when starting the app?

collisdigital commented on scripts/generate_secrets.py:
```diff
@@ -11,23 +16,73 @@ def get_file_contents(folder, filename, trim=False):
+
+
+def generate_kid_from_key(pem_data):
+    key_as_bytes = to_bytes(pem_data)
```
Not sure this is correct, I'd expect to load the PEM data using python cryptography and convert the public key to bytes, otherwise you're hashing the PEM text not the key bytes itself (see https://www.ietf.org/rfc/rfc3280.txt section 4.2.1.2)

collisdigital commented on scripts/generate_secrets.py:
```diff
@@ -11,23 +16,73 @@ def get_file_contents(folder, filename, trim=False):
+
+    pub_key = private_key.public_key()
+
+    pub_bytes = pub_key.public_bytes(Encoding.PEM, PublicFormat.SubjectPublicKeyInfo)
```
This looks better, do similar on line 27?

jonnyshaw89 replied:
Both methods create identical hex values

collisdigital replied:
ok

collisdigital commented on scripts/generate_secrets.py:
```diff
@@ -11,23 +16,73 @@ def get_file_contents(folder, filename, trim=False):
-keys = {}
+public_keys = {}
+
+# add_key_to_dict(public_keys, get_file_contents(keys_folder, 'sdc-submission-encryption-sdx-public-key.pem'))
```
Remove?

collisdigital commented on scripts/generate_secrets.py:
```diff
@@ -11,23 +16,73 @@ def get_file_contents(folder, filename, trim=False):
+    private_keys['EQ_USER_AUTHENTICATION_RRM_PRIVATE_KEY'] = get_file_contents(keys_folder, 'sdc-user-authentication-signing-rrm-private-key.pem')
+
+keys = {
+    "public_keys": public_keys,
```
May have to expand this to include key purpose

collisdigital commented on app/authentication/authenticator.py:
```diff
@@ -76,19 +78,13 @@ def store_session(metadata):
 
-    decrypted_token = decoder.decrypt_jwt_token(
-        encrypted_token,
-        current_app.config['EQ_JWT_LEEWAY_IN_SECONDS'],
```
Is the leeway still handled somewhere?

jonnyshaw89 replied:
It is now.

collisdigital commented on app/authentication/jwt_decoder.py:
```diff
@@ -1,179 +0,0 @@
-    def _check_header_values(token):
-        header = jwt.get_unverified_header(token)
-
-        if not header:
```
Being lazy, are all these checks still present somewhere else?

jonnyshaw89 replied:
Yes the library check all of this for us. the tests are still present that used to test this.

collisdigital replied:
Great

collisdigital commented on app/authentication/jwt_decoder.py:
```diff
@@ -1,179 +0,0 @@
-            if not payload:
-                raise InvalidTokenException("Missing Payload")
-            payload_decoded = payload.decode()
-            if payload_decoded == "{}":
```
And these, are they checked elsewhere now?

jonnyshaw89 replied:
All the tests have been ported across.

collisdigital replied:
:+1:

collisdigital commented on app/authentication/jwt_decoder.py:
```diff
@@ -1,179 +0,0 @@
-        header_data = json.loads(header)
-        if not header_data.get("alg"):
-            raise InvalidTokenException("Missing Algorithm")
-        if header_data.get("alg") != "RSA-OAEP":
```
And here

jonnyshaw89 replied:
accepted alg is passed into the library when decoding

collisdigital replied:
:+1:

collisdigital commented on app/cryptography/token_helper.py:
```diff
@@ -0,0 +1,97 @@
+
+        return jwe_token.payload.decode()
+    except InvalidJWEData as e:
+        raise InvalidTokenException(repr(e))
```
you can do `raise InvalidTokenException from e` which chains the exceptions together

jonnyshaw89 replied:
Thats a lot neater. The code i had was pulled from the old code.

jonnyshaw89 replied:
However your code doesn't work

collisdigital commented on app/cryptography/token_helper.py:
```diff
@@ -0,0 +1,97 @@
+    try:
+        downstream_public_jwk = secret_store.get_public_key_by_kid(purpose, kid)
+
+        _protected_header = {"alg": alg, "enc": enc, "kid": kid} if protected_header is None else protected_header
```
seems like this function is overloaded to do two things, can't this creation of a protected header be moved out so this function just takes the header?

ajmaddaford replied:
Is the protected_header arg used anywhere?

jonnyshaw89 replied:
Or remove it because i don't use arg any more

collisdigital commented on app/secrets.py:
```diff
@@ -15,8 +13,44 @@
+class SecretStore:
+    def __init__(self, secrets):
+        if 'secrets' in secrets:
+            self.secrets = secrets['secrets']
```
maybe the `if` statements should be removed and replace with `self.secrets = secrets['secrets'] or None` so at least the attribute is always set at `__init__` even if to `None` ?

collisdigital replied:
`self.secrets = secrets.get('secrets')` even

collisdigital commented on app/secrets.py:
```diff
@@ -15,8 +13,44 @@
+
+    def get_private_key_by_kid(self, purpose, kid):
+        if purpose in self.private_keys and kid in self.private_keys[purpose]:
+            return jwk.JWK.from_pem(self.private_keys[purpose][kid].encode('utf-8'), 'digitaleq'.encode('utf-8'))
```
Do we have a tech debt card about the password?

collisdigital commented on app/submitter/encrypter.py:
```diff
@@ -1,83 +1,15 @@
+def encrypt(json, secret_store):
 
-        tag = encryptor.tag
+    jwt_kid = secret_store.get_private_kids_by_purpose(PURPOSE_SUBMISSION)
```
Maybe an assert or similar to ensure we only get 1 key back?

jonnyshaw89 replied:
I might add that logic to validate_secrets so the app won't start without one.

collisdigital replied:
I was also thinking that there isn't more than 1 (there shouldn't be)

jonnyshaw89 replied:
agreed.

collisdigital commented on app/views/dev_mode.py:
```diff
@@ -6,13 +6,17 @@
 logger = get_logger()
 dev_mode_blueprint = Blueprint('dev_mode', __name__, template_folder='templates')
 
+EQ_USER_AUTHENTICATION_RRM_PRIVATE_KEY_KID = "709eb42cfee5570058ce0711f730bfbb7d4c8ade"
```
Do we have examples in the ons-schema-docs of the extract format of the kid like this, i.e. a lowercase alpha-numeric byte (0-9a-f) string exactly (?) characters long?

collisdigital replied:
Makes me think we should put a regex validator in for checking KIDs conform.

collisdigital commented on scripts/generate_secrets.py:
```diff
@@ -11,23 +19,75 @@ def get_file_contents(folder, filename, trim=False):
+    "private_keys": private_keys,
+}
 
 with open('secrets.yml', 'w') as f:
```
This file is getting more complex, should we document the structure of it somewhere, perhaps in the new design/architecture decisions log or README?

jonnyshaw89 replied:
Not the readme, already too much info in that.

collisdigital replied:
somewhere else then (top of this file??)

collisdigital commented on tests/app/authentication/__init__.py:
```diff
@@ -40,7 +40,7 @@
 
 
-TEST_DO_NOT_USE_RRM_PUBLIC_PEM = """-----BEGIN PUBLIC KEY-----
+TEST_DO_NOT_USE_UPSTREAM_PUBLIC_PEM = """-----BEGIN PUBLIC KEY-----
```
Maybe the name should include the purpose so it's obvious what it's for?

collisdigital commented on tests/app/authentication/__init__.py:
```diff
@@ -50,7 +50,7 @@
 -----END PUBLIC KEY-----"""
 
-TEST_DO_NOT_USE_RRM_PRIVATE_PEM = """-----BEGIN RSA PRIVATE KEY-----
+TEST_DO_NOT_USE_UPSTREAM_PRIVATE_PEM = """-----BEGIN RSA PRIVATE KEY-----
```
Maybe the name should include the purpose so it's obvious what it's for?

collisdigital commented on tests/app/authentication/__init__.py:
```diff
@@ -83,29 +83,32 @@
 TEST_DO_NOT_USE_PASSWORD = "digitaleq"
 
-TEST_DO_NOT_USE_RRM_PRIVATE_KEY = """-----BEGIN RSA PRIVATE KEY-----
+TEST_DO_NOT_USE_UPSTREAM_PRIVATE_KEY = """-----BEGIN RSA PRIVATE KEY-----
```
Maybe the name should include the purpose so it's obvious what it's for?

collisdigital reviewed (approved):
Looks good!

jonnyshaw89 committed:
Allow multiple encryption and signing keys to be used.
This allows for Survey Runner to support multiple upstream systems and for keys to be rotated by those systems with no downtime.
Remove private key passwords, as these were not really needed.
Added kid to the JWE header and not properly use the kid to identify which key was used when both signing and encrypting messages.

ajmaddaford commented on app/secrets.py:
```diff
@@ -15,8 +13,44 @@
+
+    def get_public_kids_by_purpose(self, purpose):
+        if purpose in self.public_keys:
+            return list(self.public_keys[purpose].keys())
```
I think using a nested structure with the kid and purpose as keys rather than values makes this code hard to understand. Flattening the keys down to something like below would make accessing keys clearer and more flexible.
```
"keys":
  - "kid": "e19091072f920cbf3ca9f436ceba309e7d814a62"
    "type": "public"
    "purpose": "authentication"
    "value": |
      -----BEGIN ENCRYPTED PRIVATE KEY-----
      ...
```

ajmaddaford replied:
I also think it might be worth considering the SecretStore returning instances of a Key class, that has properties for kid, purpose, value etc and a method `as_jwk` to return a JWK version of a key.

ajmaddaford commented on app/cryptography/token_helper.py:
```diff
@@ -0,0 +1,97 @@
+        if 'kid' in protected_header:
+            return protected_header_data['kid']
+        elif default_kid:
+            return default_kid
```
I don't think this method should know anything about a default_kid, the defaulting should be done where this method is called in `token_helper.py`

ajmaddaford commented on app/cryptography/token_helper.py:
```diff
@@ -0,0 +1,97 @@
+
+        jwe_kid = extract_kid_from_header(encrypted_token, default_kid)
+
+        sr_private_jwk = secret_store.get_private_key_by_kid(purpose, jwe_kid)
```
I think we should drop the `sr_` from this variable name

ajmaddaford commented on app/cryptography/token_helper.py:
```diff
@@ -0,0 +1,97 @@
+
+
+def encode_jwt(claims, kid, secret_store, purpose):
+    sr_private_jwk = secret_store.get_private_key_by_kid(purpose, kid)
```
I think we should drop the `sr_` from this variable name

ajmaddaford commented on app/cryptography/token_helper.py:
```diff
@@ -0,0 +1,97 @@
+    try:
+        jwt_kid = extract_kid_from_header(jwt_token)
+
+        upstream_public_jwk = secret_store.get_public_key_by_kid(purpose, jwt_kid)
```
I think we should drop the `upstream_` from this variable name

ajmaddaford commented on app/cryptography/token_helper.py:
```diff
@@ -0,0 +1,97 @@
+
+        upstream_public_jwk = secret_store.get_public_key_by_kid(purpose, jwt_kid)
+
+        _check_claims = {"exp": None, "iat": None} if check_claims is None else check_claims
```
It looks like check_claims is only overridable for the tests, I think it shouldn't be and we should re-work the tests so it's not needed.

ajmaddaford commented on app/storage/encrypted_questionnaire_storage.py:
```diff
@@ -68,11 +68,39 @@ def _generate_key(user_id, user_ik, pepper):
+            "enc": "A256GCM",
+            "kid": "1,1",
+        }
+        token = jwe.JWE(plaintext=base64url_encode(data), protected=protected_header)
```
It's probably worth calling this `jwe_token` for clarity

ajmaddaford commented on app/storage/encrypted_questionnaire_storage.py:
```diff
@@ -68,11 +68,39 @@ def _generate_key(user_id, user_ik, pepper):
+            "k": base64url_encode(self._cek),
+        }
+
+        key = jwk.JWK(**password)
```
Also probably better calling this `jwk` than `key`

ajmaddaford replied:
Might be worth factoring out creating the jwk into a method that takes the cek as a value. This method could be used by both the encrypt and decrypt methods.

ajmaddaford commented on app/storage/encrypted_questionnaire_storage.py:
```diff
@@ -68,11 +68,39 @@ def _generate_key(user_id, user_ik, pepper):
+            "k": base64url_encode(self._cek),
+        }
+
+        key = jwk.JWK(**password)
```
Also probably better calling this `jwk` than `key`

jonnyshaw89 replied:
i can't as thats what the library is called

ajmaddaford commented on app/secrets.py:
```diff
@@ -15,8 +13,44 @@
 ]
 
+PURPOSE_AUTHENTICATION = 'authentication'
+PURPOSE_SUBMISSION = 'submission'
```
I think it would be clearer in the places these are used if they were `KEY_PURPOSE_AUTHENTICATION` and `KEY_ PURPOSE_SUBMISSION`

adoublebarrel commented on scripts/generate_secrets.py:
```diff
@@ -2,6 +2,52 @@
+yaml.add_representer(LiteralUnicode, represent_literal_unicode)
+
+
+if len(sys.argv) < 2:
```
Would be nice to introduce use of argparse under a if __name__ == "__main__" block. Argparse will handle your argument validation for you and it gives a distinct visual split between class/func defs and the code that is actually doing stuff.

Also means you can import the script for unit testing, if you want to.

adoublebarrel commented on scripts/generate_secrets.py:
```diff
@@ -2,6 +2,52 @@
 import yaml
```
Worth sticking a #!/usr/bin/env python line at the top so the script can be executed from a shell without having to prefix it with "python"?

collisdigital commented:
Latest changes look good to me 👍

ajmaddaford reviewed (commented):
Looks good 👍  
I'll leave the blocking review until we're ready to deploy this breaking change to pre-prod.
